File: frontend/src/api/laundry.ts
import { apiFetch, type SessionLike } from "./client";

export type LaundryStatus = "collected" | "sent" | "received" | "closed";

export type LaundryItem = {
  id: number;
  hotel_id: number;
  batch_id: number;
  item_type: string;
  quantity: number;
  room_id?: number | null;
};

export type LaundryBatch = {
  id: number;
  hotel_id: number;
  batch_code: string;
  status: LaundryStatus;
  sent_at?: string | null;
  received_at?: string | null;
  notes?: string | null;
  created_by_user_id?: number | null;
  created_at: string;
  items: LaundryItem[];
};

export type LaundryBatchCreate = {
  notes?: string | null;
};

export type LaundryItemCreate = {
  item_type: string;
  quantity: number;
  room_id?: number | null;
};

export const listLaundryBatches = (
  filters: { statusFilter?: string; batchCode?: string } = {},
  session?: SessionLike
) => {
  const params = new URLSearchParams();
  if (filters.statusFilter) params.set("status_filter", filters.statusFilter);
  if (filters.batchCode) params.set("batch_code", filters.batchCode);
  const query = params.toString();
  return apiFetch<LaundryBatch[]>(`/api/laundry/batches${query ? `?${query}` : ""}`, { session });
};

export const createLaundryBatch = (payload: LaundryBatchCreate, session?: SessionLike) =>
  apiFetch<LaundryBatch>("/api/laundry/batches", { method: "POST", data: payload, session });

export const addLaundryItem = (batchId: number, payload: LaundryItemCreate, session?: SessionLike) =>
  apiFetch<LaundryItem>(`/api/laundry/batches/${batchId}/items`, { method: "POST", data: payload, session });

export const updateLaundryStatus = (batchId: number, status: LaundryStatus, session?: SessionLike) =>
  apiFetch<LaundryBatch>(`/api/laundry/batches/${batchId}/status`, {
    method: "PATCH",
    data: { status },
    session
  });

File: app/api/laundry.py
"""
FastAPI routes for laundry operations.
"""
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies.auth import AuthContext, require_permission
from app.services.permission_service import PERMISSION_LAUNDRY_MOVE, PERMISSION_LAUNDRY_READ
from app.services.laundry_service import (
    LaundryError,
    add_item,
    create_batch,
    get_batch,
    list_batches,
    transition_status,
)

router = APIRouter(prefix="/api/laundry", tags=["Laundry"])


class LaundryItemCreate(BaseModel):
    item_type: str
    quantity: int
    room_id: Optional[int] = None


class LaundryItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    hotel_id: int
    batch_id: int
    item_type: str
    quantity: int
    room_id: Optional[int] = None


class LaundryBatchCreate(BaseModel):
    notes: Optional[str] = None


class LaundryBatchRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    hotel_id: int
    batch_code: str
    status: str
    sent_at: Optional[datetime] = None
    received_at: Optional[datetime] = None
    notes: Optional[str] = None
    created_by_user_id: Optional[int] = None
    created_at: datetime
    items: list[LaundryItemRead] = Field(default_factory=list)


class LaundryStatusUpdate(BaseModel):
    status: str


@router.post("/batches", response_model=LaundryBatchRead, status_code=status.HTTP_201_CREATED)
def create_laundry_batch(
    data: LaundryBatchCreate,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_permission(PERMISSION_LAUNDRY_MOVE)),
):
    batch = create_batch(
        db,
        hotel_id=context.hotel_id,
        notes=data.notes,
        created_by_user_id=context.user_id,
    )
    db.commit()
    db.refresh(batch)
    return batch


@router.get("/batches", response_model=list[LaundryBatchRead])
def list_laundry_batches(
    status_filter: Optional[str] = None,
    batch_code: Optional[str] = None,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_permission(PERMISSION_LAUNDRY_READ)),
):
    return list_batches(
        db,
        hotel_id=context.hotel_id,
        filters={"status": status_filter, "batch_code": batch_code},
    )


@router.get("/batches/{batch_id}", response_model=LaundryBatchRead)
def get_laundry_batch(
    batch_id: int,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_permission(PERMISSION_LAUNDRY_READ)),
):
    try:
        return get_batch(db, hotel_id=context.hotel_id, batch_id=batch_id)
    except LaundryError as exc:
        raise HTTPException(status_code=404, detail=str(exc))


@router.post("/batches/{batch_id}/items", response_model=LaundryItemRead, status_code=status.HTTP_201_CREATED)
def add_laundry_item(
    batch_id: int,
    data: LaundryItemCreate,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_permission(PERMISSION_LAUNDRY_MOVE)),
):
    try:
        item = add_item(db, hotel_id=context.hotel_id, batch_id=batch_id, **data.model_dump())
    except LaundryError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    db.commit()
    db.refresh(item)
    return item


@router.patch("/batches/{batch_id}/status", response_model=LaundryBatchRead)
def update_laundry_status(
    batch_id: int,
    data: LaundryStatusUpdate,
    db: Session = Depends(get_db),
    context: AuthContext = Depends(require_permission(PERMISSION_LAUNDRY_MOVE)),
):
    try:
        batch = transition_status(db, hotel_id=context.hotel_id, batch_id=batch_id, new_status=data.status)
    except LaundryError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    db.commit()
    db.refresh(batch)
    return batch
